### ISSUE TITLE
Don't compile C/C++ header (.h) files

### ISSUE THREAD

**Issue by clevengr** [pc2ccs/pc2v9#238]
**Describe the issue**: 

This bug only occurs when more than one file can be submitted by a team.   This happens when
in Shadow Mode, on the pc2 team fat client and directly via the PC^2 Java API.    WTI does not
support submitting additional files (See #242)

During the NADC 21 Practice 2 contest, several teams submitted C++ solutions made up of multiple files (which is legal).  Some of the files were typical ".h" files.   The execution command for C++ was `g++ -g -O2 -std=gnu++17 -static {files}`.  This results in substituting for `{files}` the list of all submitted files -- including the ".h" files.  This in turn tells the compiler it should _compile_ those ".h" files -- which results in a compiler error (".h" files get _included_ in the compilation of .c/.cpp files which _reference_ them).

The substitution for {files} needs to be smart enough to recognize that it should only substitute _source_ files, not "include" (.h) files.

**To Reproduce**: 
- Define C++ as a language using the following:  `g++ -g -O2 -std=gnu++17 -static {files}`
- Submit a program consisting of two files: a main.cpp file and a separate ".h" header file.  Make sure the main file contains valid C++ code and references the header file with a `#include` statement.  
- Execute the program on  a machine with g++ installed.

**Expected behavior**: 
The program should compile successfully and should execute (doing whatever the C++ code says).

**Actual behavior**: 
The program fails to compile.

**Do you have any specific suggestions for how your feature would be ***implemented*** in PC^2?** 

Identify a list of allowed extensions per Language.
Change the {files} substitution to only include include files with allowed extensions

C++ source extensions: cpp, C, c++

**Additional Info**:
An example of this which happened in the NADC21 Practice 2 contest was where a team submitted a C++ program with two files: `template.cpp` and `SIMD.h`.  The contents of SIMD.h was:
```
/**
 * Author: Simon Lindholm
 * Date: 2015-03-18
 * License: CC0
 * Source: https://software.intel.com/sites/landingpage/IntrinsicsGuide/
 * Description: Cheat sheet of SSE/AVX intrinsics, for doing arithmetic on several numbers at once.
 * Can provide a constant factor improvement of about 4, orthogonal to loop unrolling.
 * Operations follow the pattern \texttt{"\_mm(256)?\_name\_(si(128|256)|epi(8|16|32|64)|pd|ps)"}. Not all are described here;
 * grep for \texttt{\_mm\_} in \texttt{/usr/lib/gcc/{*}/4.9/include/} for more.
 * If AVX is unsupported, try 128-bit operations, "emmintrin.h" and \#define \texttt{\_\_SSE\_\_} and \texttt{\_\_MMX\_\_} before including it.
 * For aligned memory use \texttt{\_mm\_malloc(size, 32)} or \texttt{int buf[N] alignas(32)}, but prefer loadu/storeu.
 */
#pragma once

#pragma GCC target ("avx2") // or sse4.1
#include "immintrin.h" /** keep-include */

typedef __m256i mi;
#define L(x) _mm256_loadu_si256((mi*)&(x))

// High-level/specific methods:
// load(u)?_si256, store(u)?_si256, setzero_si256, _mm_malloc
// blendv_(epi8|ps|pd) (z?y:x), movemask_epi8 (hibits of bytes)
// i32gather_epi32(addr, x, 4): map addr[] over 32-b parts of x
// sad_epu8: sum of absolute differences of u8, outputs 4xi64
// maddubs_epi16: dot product of unsigned i7's, outputs 16xi15
// madd_epi16: dot product of signed i16's, outputs 8xi32
// extractf128_si256(, i) (256->128), cvtsi128_si32 (128->lo32)
// permute2f128_si256(x,x,1) swaps 128-bit lanes
// shuffle_epi32(x, 3*64+2*16+1*4+0) == x for each lane
// shuffle_epi8(x, y) takes a vector instead of an imm

// Methods that work with most data types (append e.g. _epi32):
// set1, blend (i8?x:y), add, adds (sat.), mullo, sub, and/or,
// andnot, abs, min, max, sign(1,x), cmp(gt|eq), unpack(lo|hi)

int sumi32(mi m) { union {int v[8]; mi m;} u; u.m = m;
	int ret = 0; rep(i,0,8) ret += u.v[i]; return ret; }
mi zero() { return _mm256_setzero_si256(); }
mi one() { return _mm256_set1_epi32(-1); }
bool all_zero(mi m) { return _mm256_testz_si256(m, m); }
bool all_one(mi m) { return _mm256_testc_si256(m, one()); }

ll example_filteredDotProduct(int n, short* a, short* b) {
	int i = 0; ll r = 0;
	mi zero = _mm256_setzero_si256(), acc = zero;
	while (i + 16 <= n) {
		mi va = L(a[i]), vb = L(b[i]); i += 16;
		va = _mm256_and_si256(_mm256_cmpgt_epi16(vb, va), va);
		mi vp = _mm256_madd_epi16(va, vb);
		acc = _mm256_add_epi64(_mm256_unpacklo_epi32(vp, zero),
			_mm256_add_epi64(acc, _mm256_unpackhi_epi32(vp, zero)));
	}
	union {ll v[4]; mi m;} u; u.m = acc; rep(i,0,4) r += u.v[i];
	for (;i<n;++i) if (a[i] < b[i]) r += a[i]*b[i]; // <- equiv
	return r;
}
```
The contents of template.cpp was:
```
#include <bits/stdc++.h>
using namespace std;

#define rep(i, a, b) for(int i = a; i < (b); ++i)
#define all(x) begin(x), end(x)
#define sz(x) (int)(x).size()
typedef long long ll;
typedef pair<int, int> pii;
typedef vector<int> vi;

#include "SIMD.h"

int main() {
	cin.tie(0)->sync_with_stdio(0);
	cin.exceptions(cin.failbit);
}
```

This is a legal C++ program which compiles correctly and produces an `a.out` file when compiled from the command line.  However, when run within PC2 it generates a compiler error.   The compiler output message was:
`Compilation error when compiling program "SIMD.h" using the command "g++ -g -O2 -std=gnu++17 -static SIMD.h template.cpp"`

**Comment by lane55** [comment]
The {files} as stated above value is currently all files submitted by the team.
The {files} substitution will be changed to only include the main file
and any remaining files that have the extensions (end with): .C, .cpp, or .c

**Comment by lane55** [comment]
Upon further thought.   Rather than use a set of files extenstions to allow/include,
it makes more sense to remove files with certain extensioins, the "include" file extension.

Those extensions are: .h, .hh,  and .H

Note that the main file is always included in the {files} value, so if a team selects .h as their main file, oddness
will ensue.

**Comment by lane55** [comment]
This may be solved by #238

**Comment by clevengr** [comment]
> `This may be solved by https://github.com/pc2ccs/pc2v9/issues/238`

I do not understand this comment.  This is a comment made in (contained in) Issue #238.  What does it mean to say "This may be solved by #238"?

**Comment by lane55** [comment]
Marked as Analysis Required to also review if may be solved by #238

**Comment by troy2914** [comment]
This was fixed per John Buck.
